### PR TITLE
feat(objective): pre-select lasso paths

### LINKED ISSUE #9: feat: use better paths for the solvers
Many solvers for fitting the full path, such as glmnet, employ path-stopping criteria to stop the path prematurely if the model saturates before completion. This is, in my opinion, very reasonable behavior since otherwise we an run into weird degenerate solutions, but we've had to turn it off because otherwise the algorithms cannot be compared properly.

One way to fix this would be to precompute a reasonable path for each data set using a solver that employs these types of checks (such as glmnet or Lasso.jl) and store the path along with each data set, but I'm not sure if that would work with the current API.

### PR DESCRIPTION
**Pull request by jolars**
Fix #9 by fitting the full path using glmnet's path-stopping criteria
with the sklearn solver once for each data set.

### CODE REVIEW

File: objective.py [hunk 1]
```diff
@@ -26,29 +29,79 @@ def set_data(self, X, y):
...
 
+        # the following code fits the path once using the same path-stopping
+        # criteria that glmnet uses to find a reasonable lambda path in order
+        # to obtain a reasonable lambda sequence
         if self.lambda_min_ratio is None:
             self.lambda_min_ratio = 1e-2 if self.n_samples < self.n_features else 1e-4
 
         lambda_max = self._get_lambda_max()
-        self.lambdas = np.logspace(
+        lambdas = np.logspace(
```
Comment (mathurinm) [inline objective.py]: I think it's clearer to use np.geomspace and it should be equivalent, right ?

File: objective.py [hunk 2]
```diff
@@ -1,7 +1,10 @@
 import numpy as np
-from benchopt import BaseObjective
+from benchopt import BaseObjective, safe_import_context
 from numpy.linalg import norm
```
Comment (mathurinm) [inline objective.py]: delay to import_ctx for faster autocompletion

File: objective.py [hunk 3]
```diff
@@ -26,29 +29,79 @@ def set_data(self, X, y):
...
+            tol=1e-4,
+            fit_intercept=self.fit_intercept,
+            warm_start=True,
+        )
+
+        r = y - np.mean(y) if self.fit_intercept else y.copy()
+        null_dev = 0.5 * np.linalg.norm(r) ** 2
+        dev_prev = null_dev
+
+        i = 0
```
Comment (mathurinm) [inline objective.py:61]: can be removed

File: objective.py [hunk 4]
```diff
@@ -26,29 +29,79 @@ def set_data(self, X, y):
...
+        ever_active = np.array([])
+
+        for i, lam in enumerate(lambdas):
+            lasso.alpha = lam / self.n_samples
+            lasso.fit(X, y)
+
+            w = lasso.coef_
+            intercept = lasso.intercept_
+
+            dev = 0.5 * np.linalg.norm(y - X @ w - intercept) ** 2
```
Comment (mathurinm) [inline objective.py:71]: if you like it, you can use clf.predict(X)  (to avoid defining w and intercept

File: objective.py [hunk 5]
```diff
@@ -26,29 +29,79 @@ def set_data(self, X, y):
...
+
+            if (
+                n_nonzero >= dfmax
+                or dev_ratio > 0.999
+                or (i > 0 and dev_change < 1e-5)
+                or len(ever_active) >= pmax
+            ):
+                break
+
+        self.lambdas = lambdas[: i + 1]
```
Comment (mathurinm) [inline objective.py:88]: no space after : (?)

File: objective.py (hunk 4 above)
Comment (mathurinm) [reply objective.py:71]: ah you need w later

**Review by mathurinm — APPROVED** [review]
LGTM, minor comments

File: objective.py (hunk 5 above)
Comment (jolars) [reply objective.py:88]: > LGTM, minor comments

Thanks for the review and useful comments!

File: objective.py (hunk 1 above)
Comment (jolars) [reply objective.py]: You're right, thanks, I thought geomspace would be with the 10 base too because logspace is.

File: objective.py (hunk 3 above)
Comment (jolars) [reply objective.py:61]: Yes, it's just there to please the linter.

Comment (mathurinm) [reply objective.py:61]: the linter complains because if lambdas is empty i is not initialized ?

Comment (jolars) [reply objective.py:61]: Yes, which I know is a false positive. I should just ignore it, but I guess I'm just too obedient to do so :(

**Pull request merged**